lib/project_types/script/layers/infrastructure/script_service.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require "base64"
+require "json"
+
+module Script
+  module Layers
+    module Infrastructure
+      class ScriptService
+        include SmartProperties
+        property! :ctx, accepts: ShopifyCli::Context
+
+        def deploy(
+          extension_point_type:,
+          schema:,
+          script_name:,
+          script_content:,
+          compiled_type:,
+          api_key: nil,
+          force: false
+        )
+          query_name = "app_script_update_or_create"
+          variables = {
+            extensionPointName: extension_point_type.upcase,
+            title: script_name,
+            sourceCode: Base64.encode64(script_content),
+            language: compiled_type,
+            schema: schema,
+            force: force,
+          }
+          resp_hash = script_service_request(query_name: query_name, api_key: api_key, variables: variables)
+          user_errors = resp_hash["data"]["appScriptUpdateOrCreate"]["userErrors"]
+
+          return resp_hash if user_errors.empty?
+
+          if user_errors.any? { |e| e['tag'] == 'already_exists_error' }
+            raise Errors::ScriptRedeployError, api_key
+          else
+            raise Errors::ScriptServiceUserError.new(query_name, user_errors.to_s)
+          end
+        end
+
+        private
+
+        class ScriptServiceAPI < ShopifyCli::API
+          property(:api_key, accepts: String)
+          property(:shop_id, accepts: Integer)
+
+          def self.query(ctx, query_name, api_key: nil, shop_domain: nil, variables: {})
+            api_client(ctx, api_key, shop_domain).query(query_name, variables: variables)
+          end
+
+          def self.api_client(ctx, api_key, shop_domain)
+            new(
+              ctx: ctx,
+              url: 'https://script-service.myshopify.io/graphql',
+              token: '',
+              api_key: api_key,
+              shop_id: shop_domain&.to_i
+            )
+          end
+
+          def auth_headers(*)
+            tokens = { "APP_KEY" => api_key, "SHOP_ID" => shop_id }.compact.to_json
+            { "X-Shopify-Authenticated-Tokens" => tokens }
+          end
+        end
+        private_constant(:ScriptServiceAPI)
+
+        class PartnersProxyAPI < ShopifyCli::PartnersAPI
+          def query(query_name, variables: {})
+            variables[:query] = load_query(query_name)
```

Comment:
Why did you have to do this?

Reply:
We make a request to partners which validates that the partner can permission to access the app/shop it wants to. This is general so that we can have multiple mutations with the same validations. If the partner has permission, then we make an authenticated request in Partners to our backend service to perform this query. So we pass the query as a variable so Partner's knows what to do